rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,155 @@
+
+The idea here is to:
+- expose host-side scalars through a new  memory descriptor create function
+- have user pass a memory with cpu engine to execute function.
+
+The new kind of memory descriptor would be created through a dedicated function
+It would be very lightweights and be limited:
+- only scalar value, so no `ndims`, no `dims`, ...
+- only scalar datatype can be configurable by user.
+
+```C
+dnnl_status_t DNNL_API dnnl_memory_desc_create_host_scalar(
+        dnnl_memory_desc_t *memory_desc, dnnl_data_type_t data_type);
+```
+
+Internal considerations: we would use a new `format_kind` with no
+associated `format_desc`. This would avoid aliasing host-side scalars
+with blocked descriptors, and hence host-side scalars will need to be
+explicitely supported by implementations.
+
+For users passing primitive objects around and querying them, it makes
+sense to expose this format kind, so users can query if a given input
+is a host-side scalar, and pass it accordingly to primitive execution.
+
+
+Pros:
+- avoids user from recreating primitives when scalars are changed
+- simple enabling internally
+
+Cons:
+- ties the memory descriptor to a specific device (host) which we
+  avoided so far.
+
+
+### Option 3: explicitely take engine kind in memory desc.
+
+Here, we would allow users to pass any memory on host or device.
+This would officially tie a given memory descritor to a device kind.
+
+For compatibility reasons, we could either use `undef` to specify that
+a memory desc will be ties to the same engine kind as the
+implementation, or introduce a new `engine_kind` for that purpose.
+
+A few points need to be made:
+- mixing host and device memory is already possible through usm shared
+  memory for both sycl and opencl runtime.
+- there is no clear benefit to mix host and device memory, except for
+  the case where the value to pass is a scalar, in which case it can
+  be passed to device kernel as a parameter (instead of passing a
+  pointer).
+- even if two memories share the same engine kind, it does not
+  guarentee that they will both be accessible from the same context
+  (they might be tied to different engine objects).
+
+As a result, we don't recommend to specify engine kind for memory
+descriptors as it would likely provide little benefit over host-side
+scalar specific memory descriptor, but being more general, it would
+also come with more complications.
+
+Pros:
+- clear semantic: all memory descritors are tied to an engine kind
+
+Cons:
+- full buffers sharing across devices can already be achieved through runtime mecanisms.
+- no clear benefit for performance other than specific scalar case.
+
+## How should host-side scalar be passed to execute call?
+
+Currently, the oneDNN primitive API only accepts `dnnl::memory` objects
+(through an `unordered_map` in C++ interface, or an array of
+`arg_kind`/`memory` for C interface).
+
+We have two paths:
+- encapsulate host-side scalars in memory objects: this would create
```

Comment:
Creating a memory object for a host scalar requires to create a host engine as well:
- when `CPU_RT=NONE`, I'm not sure if the user can create a CPU engine.
- when `CPU_RT=SYCL`, the address of the host scalar is not a SYCL memory.

Or, do you think we can use a GPU engine to create a memory object for a host scalar?

I think we discussed it when adding the host scalar support to graph API, but feel like to also mention it here.

Reply:
Resolving this thread for now, as RFC got updated and this doesn't seem to be relevant anymore.. Please feel free to reopen / start a new one